### PR TITLE
Update ScaffoldColorPicker to use Spectrum.

### LINKED ISSUE #2014: Replace colour picker widget with something better supported
Possibilities:
- [CSS Script (basic)](https://www.cssscript.com/demo/color-picker-component-pickr/)
- [CSS Script (alpha)](https://www.cssscript.com/demo/color-picker-alpha-selection/)
- [Spectrum](http://bgrins.github.io/spectrum/)
- [Jscolor](http://jscolor.com/)

### PR DESCRIPTION
**Pull request by taylortom**
Fixes #2014 (includes and extends #2017).

## Notes
- The old `ScaffoldColorPicker` has been renamed to the **GB-EN** equivalent: `ScaffoldColourPicker`
- Improvements/amendments over Spectrum plugin defaults:
  * Styling has been updated to more closely match AT theme
  * Trash, apply and cancel buttons have been removed
  * Page scrolling is disabled when colour picker popup is opened

## Default Spectrum input
screen shot 2018-07-23 at 17 43 37

## Customised Spectrum input
screen shot 2018-08-06 at 16 17 22

### CODE REVIEW

File: frontend/src/modules/scaffold/views/scaffoldColorPickerView.js [hunk 1]
```diff
@@ -20,20 +20,22 @@ define([
...
       this.$el.spectrum({
         color: this.value,
         showInput: true,
         showAlpha: true,
-        allowEmpty: true,
         showInitial: true,
         showInput: true,
         showPalette: true,
+        showButtons: true,
+        cancelText: "Cancel selection",
```
Comment (tomgreenfield) [inline frontend/src/modules/scaffold/views/scaffoldColorPickerView.js]: Needs localising and could do with `font-family: @font-family;`

**Comment by lc-thomasberger** [comment]
works for me

**Comment by lc-thomasberger** [comment]
When I test this I get the following error when trying to edit a page with the editable theme selected:
```
Uncaught TypeError: constructorFn is not a constructor
    at child.createEditor (backbone-forms.js:842)
    at child.initialize (backbone-forms.js:803)
    at child.Backbone.View (backbone.js:991)
    at child [as constructor] (backbone.js:1541)
    at new child (backbone.js:1541)
    at child.createField (backbone-forms.js:136)
    at child.<anonymous> (backbone-forms.js:83)
    at Array.forEach (<anonymous>)
    at Function._.each._.forEach (underscore.js:79)
    at child.initialize (backbone-forms.js:81)
createEditor @ backbone-forms.js:842
initialize @ backbone-forms.js:803
Backbone.View @ backbone.js:991
child @ backbone.js:1541
child @ backbone.js:1541
createField @ backbone-forms.js:136
(anonymous) @ backbone-forms.js:83
_.each._.forEach @ underscore.js:79
initialize @ backbone-forms.js:81
Backbone.View @ backbone.js:991
child @ backbone.js:1541
render @ backbone-forms.js:2100
(anonymous) @ backbone-forms.js:932
each @ jquery.js:384
each @ jquery.js:136
render @ backbone-forms.js:926
(anonymous) @ backbone-forms.js:736
_.each._.forEach @ underscore.js:87
(anonymous) @ backbone-forms.js:735
each @ jquery.js:384
each @ jquery.js:136
render @ backbone-forms.js:729
(anonymous) @ backbone-forms.js:246
_.each._.forEach @ underscore.js:79
(anonymous) @ backbone-forms.js:245
each @ jquery.js:384
each @ jquery.js:136
render @ backbone-forms.js:239
render @ backbone-forms.js:2109
(anonymous) @ backbone-forms.js:932
each @ jquery.js:384
each @ jquery.js:136
render @ backbone-forms.js:926
(anonymous) @ backbone-forms.js:736
_.each._.forEach @ underscore.js:87
(anonymous) @ backbone-forms.js:735
each @ jquery.js:384
each @ jquery.js:136
render @ backbone-forms.js:729
(anonymous) @ backbone-forms.js:246
_.each._.forEach @ underscore.js:79
(anonymous) @ backbone-forms.js:245
each @ jquery.js:384
each @ jquery.js:136
render @ backbone-forms.js:239
Scaffold.buildForm @ index.js:240
renderContentObjectEdit @ index.js:34
route @ index.js:17
success @ index.js:27
options.success @ backbone.js:433
fire @ jquery.js:3143
fireWith @ jquery.js:3255
done @ jquery.js:9309
callback @ jquery.js:9713
XMLHttpRequest.send (async)
send @ jquery.js:9659
ajax @ jquery.js:9210
Backbone.ajax @ backbone.js:1197
Backbone.sync @ backbone.js:1178
sync @ backbone.js:284
fetch @ backbone.js:437
(anonymous) @ index.js:24
triggerEvents @ backbone.js:205
trigger @ backbone.js:146
(anonymous) @ origin.js:50
callTap @ origin.js:109
callTaps @ origin.js:112
trigger @ origin.js:46
triggerEvent @ index.js:50
done @ editorDataLoader.js:112
triggerEvents @ backbone.js:204
trigger @ backbone.js:146
(anonymous) @ origin.js:50
callTap @ origin.js:109
callTaps @ origin.js:112
trigger @ origin.js:46
(anonymous) @ editorDataLoader.js:95
success @ editorDataLoader.js:177
options.success @ backbone.js:433
fire @ jquery.js:3143
fireWith @ jquery.js:3255
done @ jquery.js:9309
callback @ jquery.js:9713
XMLHttpRequest.send (async)
send @ jquery.js:9659
ajax @ jquery.js:9210
Backbone.ajax @ backbone.js:1197
Backbone.sync @ backbone.js:1178
sync @ configModel.js:22
fetch @ backbone.js:437
fetchEditorData @ editorDataLoader.js:174
loadCourseData @ editorDataLoader.js:92
(anonymous) @ editorDataLoader.js:76
triggerEvents @ backbone.js:204
trigger @ backbone.js:146
(anonymous) @ origin.js:50
callTap @ origin.js:109
callTaps @ origin.js:112
trigger @ origin.js:46
(anonymous) @ editorDataLoader.js:59
success @ editorDataLoader.js:177
options.success @ backbone.js:861
fire @ jquery.js:3143
fireWith @ jquery.js:3255
done @ jquery.js:9309
callback @ jquery.js:9713
XMLHttpRequest.send (async)
send @ jquery.js:9659
ajax @ jquery.js:9210
Backbone.ajax @ backbone.js:1197
Backbone.sync @ backbone.js:1178
sync @ backbone.js:627
fetch @ backbone.js:865
fetchEditorData @ editorDataLoader.js:174
loadGlobalData @ editorDataLoader.js:57
triggerEvents @ backbone.js:204
trigger @ backbone.js:146
(anonymous) @ origin.js:50
callTap @ origin.js:109
callTaps @ origin.js:112
trigger @ origin.js:46
(anonymous) @ origin.js:15
success @ schemasModel.js:20
options.success @ backbone.js:433
fire @ jquery.js:3143
fireWith @ jquery.js:3255
done @ jquery.js:9309
callback @ jquery.js:9713
XMLHttpRequest.send (async)
send @ jquery.js:9659
ajax @ jquery.js:9210
Backbone.ajax @ backbone.js:1197
Backbone.sync @ backbone.js:1178
sync @ backbone.js:284
fetch @ backbone.js:437
loadData @ schemasModel.js:17
triggerEvents @ backbone.js:205
trigger @ backbone.js:146
(anonymous) @ origin.js:50
callTap @ origin.js:109
callTaps @ origin.js:112
trigger @ origin.js:46
(anonymous) @ origin.js:14
(anonymous) @ underscore.js:754
success @ origin.js:29
options.success @ backbone.js:433
fire @ jquery.js:3143
fireWith @ jquery.js:3255
done @ jquery.js:9309
callback @ jquery.js:9713
XMLHttpRequest.send (async)
send @ jquery.js:9659
ajax @ jquery.js:9210
Backbone.ajax @ backbone.js:1197
Backbone.sync @ backbone.js:1178
sync @ backbone.js:284
fetch @ backbone.js:437
startSession @ origin.js:26
(anonymous) @ app.js:73
execCb @ require.js:29
check @ require.js:18
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
enable @ require.js:23
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
(anonymous) @ require.js:26
setTimeout (async)
j.nextTick @ require.js:32
h @ require.js:26
requirejs @ require.js:31
(anonymous) @ app.js:72
execCb @ require.js:29
check @ require.js:18
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
enable @ require.js:23
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
(anonymous) @ require.js:26
setTimeout (async)
j.nextTick @ require.js:32
h @ require.js:26
requirejs @ require.js:31
loadPlugins @ app.js:58
(anonymous) @ app.js:69
execCb @ require.js:29
check @ require.js:18
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
enable @ require.js:23
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
(anonymous) @ require.js:17
(anonymous) @ require.js:8
execCb @ require.js:29
check @ require.js:18
(anonymous) @ require.js:23
(anonymous) @ require.js:8
(anonymous) @ require.js:23
y @ require.js:7
emit @ require.js:23
check @ require.js:19
enable @ require.js:23
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
load (async)
j.load @ require.js:33
load @ require.js:29
load @ require.js:18
(anonymous) @ require.js:17
(anonymous) @ require.js:8
execCb @ require.js:29
check @ require.js:18
enable @ require.js:23
init @ require.js:17
(anonymous) @ require.js:26
setTimeout (async)
j.nextTick @ require.js:32
h @ require.js:26
fetch @ require.js:17
check @ require.js:20
enable @ require.js:23
enable @ require.js:27
(anonymous) @ require.js:23
(anonymous) @ require.js:8
y @ require.js:7
enable @ require.js:22
init @ require.js:17
D @ require.js:14
completeLoad @ require.js:28
onScriptLoad @ require.js:29
Show 548 more blackboxed frames
ckeditor.js:306 Uncaught TypeError: Cannot read property 'unselectable' of null
    at b (ckeditor.js:306)
    at a.<anonymous> (ckeditor.js:303)
    at a.h (ckeditor.js:10)
    at a.<anonymous> (ckeditor.js:11)
    at a.window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.editor.CKEDITOR.editor.fire (ckeditor.js:13)
    at a.fireOnce (ckeditor.js:12)
    at a.window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.editor.CKEDITOR.editor.fireOnce (ckeditor.js:13)
    at Object.<anonymous> (ckeditor.js:234)
    at g (ckeditor.js:215)
    at Object.load (ckeditor.js:216)
```

**Comment by taylortom** [comment]
@lc-thomasberger I think that's related to the renaming of the `ColorPicker` input type to `ColourPicker`. The latest version of my [community theme](https://github.com/taylortom/adapt-theme-community) should work.

**Comment by tomgreenfield** [comment]
* Suggest `allowEmpty: true` to try and support empty strings in the field
* Suggest we change ticks on the palette colours to borders
  * Not too sure how useful the palette is in general, if you'd prefer to remove it

**Comment by tomgreenfield** [comment]
As requested, some example styles to go [here](https://github.com/adaptlearning/adapt_authoring/pull/2032/files#diff-96142c599bf2e37ed71a8482724f5f0aR22) for the ticks → border bit (could do with tweaking):
```less
border-color: #eff7f9;
box-sizing: border-box;
&.sp-thumb-active {
  border: 2px solid #000;
}
.sp-thumb-inner {
  border: 1px solid #eff7f9;
  border-radius: 50%;
  background-image: none;
}
```

**Comment by taylortom** [comment]
@tomgreenfield have added the new palette styling as discussed

**Comment by tomgreenfield** [comment]
Very nice. Now just need to fix resetting to an empty default `""`.

**Comment by taylortom** [comment]
Done, have enabled `allowEmpty`, and hidden the button via LESS.

**Pull request merged**